Support for auto-mitigate for Monitor scheduled query rules alerts #13036

cloutier opened the pull request.

cloutier commented:
I've incorporated changes from feedback from @Klaas- and I'm ready for another round of review

Klaas- commented:
I think you can use "ConflictsWith" to ensure that no one sets them both, right?

cloutier commented:
"ConflictsWith" is interesting! I was looking at other ressources like application_gateway and they were validating in the update function, but ConflictsWith is the prefered way, right? Is it somewhat new?

Klaas- commented:
I am not sure, I just started with terraform today :)

cloutier commented:
Ah no problem! @katbyte and @tombuildsstuff, what do you think?

Klaas- commented:
So I tested the way it's currently implemented in this PR. 
The error when using thorttling and auto_mitigate together, only in apply the error will happen? Is that wanted behavior?

2nd problem: even if I leave throttling away and I test it I still get `Status=400 Code="BadRequest" Message="Auto mitigation must be disabled when action suppression is set"`.
Looking at the debug it still sends `"throttlingInMin":0`.
I am guessing they need to be not send at all for that API to work -- or would this be considered an API error on microsofts part because 0 minutes equals turned off?

Klaas- commented:
there is another issue: https://github.com/hashicorp/terraform-provider-azurerm/issues/12341 that says this will need a new azure go sdk version

Klaas- commented:
I am not sure where the throttlingInMin must be suppressed when it's zero, would that happen somewhere here:
https://github.com/hashicorp/terraform-provider-azurerm/blob/eb2c5e8c45978537e7d32a2f12494ee5b6148994/internal/services/monitor/monitor_scheduled_query_rules_alert_resource.go#L409

or should that be handled here:
https://github.com/Azure/azure-sdk-for-go/blob/fbb66cedeb94af0a56ff6e8f31a44314af3ef114/services/preview/monitor/mgmt/2021-07-01-preview/insights/models.go#L479-L481

what I tested is inside the models.go I can just check if it's bigger than 0 and it works as expected.

I do not think it still needs a newer go sdk version, as stated in #12341, maybe that's needed for the 1 minute intervals not the statefulness.

cloutier commented:
I've done changes according to feedback. I've switched to use ConflictsWith and simplified the Update function.

Ideally I would touch as little as possible "throttle" in this PR. I want to help out a coworker that needs "auto_mitigate" as soon as possible. 

I'm ready for another round of review

Klaas- commented:
@cloutier did you see https://github.com/hashicorp/terraform-provider-azurerm/pull/13213#issuecomment-912552646 I think you also need to change something in https://github.com/hashicorp/terraform-provider-azurerm/blob/bf33fce101086846587047c13ac19b66a182910f/internal/services/monitor/monitor_scheduled_query_rules_alert_resource.go#L406 otherwise it will always send `"throttlingInMin":0`

I am not sure if you should change something in azure sdk or here though :)

cloutier commented:
@Klaas- Yes I have! This is outside the scope of what I want to do with this PR though

Klaas- commented:
but without that auto mitigate won't work

Klaas- commented:
any request you send to the azure log alert api with "throttlingInMin":0 (ie no throttling defined in terraform) where auto mitigration is true will lead to `Status=400 Code="BadRequest" Message="Auto mitigation must be disabled when action suppression is set"`

katbyte commented on internal/services/monitor/monitor_scheduled_query_rules_alert_resource.go:
```diff
@@ -92,6 +92,12 @@ func resourceMonitorScheduledQueryRulesAlert() *pluginsdk.Resource {
 				Required:     true,
 				ValidateFunc: azure.ValidateResourceID,
 			},
+			"auto_mitigate": {
```
in keeping with booleans elsewhere in the provider can we change this to
```suggestion
			"auto_mitigation_enabled": {
```

katbyte reviewed (changes requested):
Thanks @cloutier - overall this looks great, i have one suggestion about the schema and could we add the property to the complete test to ensure it works? thanks!

cloutier commented:
Thanks for the feedback @katbyte! I've done the changes you suggested. 

I've added a test for the undocumented quirk that @Klaas- was talking about, but I saw it the logs of the pipeline that it hasn't actually run: 
```
     testcase.go:88: Acceptance tests skipped unless env 'TF_ACC' set
--- SKIP: TestAccMonitorScheduledQueryRules_AutoMitigate (0.00s)
```

When is that test actually run? Also should I do things locally to run the test or is there a way to run it in this PR verification?

Klaas- commented:
@cloutier I don't think the PR tests actually do tests on azure, if I understood https://github.com/hashicorp/terraform-provider-azurerm/blob/main/README.md#developing-the-provider correctly you need to run those manually.
for the tests I ran last friday I simply compiled your branch myself and put it into my local terraform also as described in the readme. I'll do the same tests with your updates next week (I'm off this week, and I don't have a private azure account). But as I said earlier, I don't think the changes suffice because you need to not send throttlingInMin at all which is not the case currently because the unset int32 is 0 and not nil.

cloutier commented:
I think I've finally figured a way to not send throttling if it's not set (in the last commit). 

My question for tests is more: Does it run once upon merges? Or just before releases? Or from time to time?

Klaas- commented on internal/services/monitor/monitor_scheduled_query_rules_alert_resource.go:
```diff
@@ -395,17 +396,20 @@ func expandMonitorScheduledQueryRulesAlertingAction(d *pluginsdk.ResourceData) *
+		OdataType:  insights.OdataTypeBasicActionOdataTypeMicrosoftWindowsAzureManagementMonitoringAlertsModelsMicrosoftAppInsightsNexusDataContractsResourcesScheduledQueryRulesAlertingAction,
+	}
+
+	if throttlingOk {
```
maybe this can happen simpler, can we just check for `throttling > 0` here? then you don't need to change the Schema and you can remove throttlingOk

stephybun replied:
We could handle the default case where throttling is 0 here instead of setting the default to `nil` in the schema.
```suggestion
	if throttling, ok := d.Get("throttling").(int); ok && throttling != 0 {
		action.ThrottlingInMin = utils.Int32(int32(throttling))
	}
```

stephybun commented on internal/services/monitor/monitor_scheduled_query_rules_alert_resource.go:
```diff
@@ -92,6 +92,12 @@ func resourceMonitorScheduledQueryRulesAlert() *pluginsdk.Resource {
+			"auto_mitigation_enabled": {
+				Type:          pluginsdk.TypeBool,
+				Optional:      true,
+				Default:       true,
```
Set default to false here, otherwise when throttling is specified instead, autoMitigate is still always sent along as true in the REST call
```suggestion
				Default:       false
```

stephybun commented on internal/services/monitor/monitor_scheduled_query_rules_alert_resource.go:
```diff
@@ -126,9 +132,11 @@ func resourceMonitorScheduledQueryRulesAlert() *pluginsdk.Resource {
-				ValidateFunc: validation.IntBetween(0, 10000),
+				Type:          pluginsdk.TypeInt,
+				Optional:      true,
+				Default:       nil,
```
We shouldn't need this Default value anymore.
```suggestion
```

stephybun commented on internal/services/monitor/monitor_scheduled_query_rules_alert_resource.go:
```diff
@@ -257,11 +266,12 @@ func resourceMonitorScheduledQueryRulesAlertCreateUpdate(d *pluginsdk.ResourceDa
+			Source:       source,
+			Schedule:     schedule,
+			Action:       action,
+			AutoMitigate: &autoMitigate,
```
```suggestion
			AutoMitigate: utils.Bool(autoMitigate),
```

stephybun commented on internal/services/monitor/monitor_scheduled_query_rules_alert_resource.go:
```diff
@@ -385,17 +396,20 @@ func expandMonitorScheduledQueryRulesAlertingAction(d *pluginsdk.ResourceData) *
 	severityRaw := d.Get("severity").(int)
 	severity := strconv.Itoa(severityRaw)
-	throttling := d.Get("throttling").(int)
+	throttling, throttlingOk := d.GetOk("throttling")
```
This get and the if conditional below could be compressed together - see my comment for line 411

stephybun commented on internal/services/monitor/monitor_scheduled_query_rules_alert_resource_test.go:
```diff
@@ -85,6 +86,75 @@ func TestAccMonitorScheduledQueryRules_AlertingActionCrossResource(t *testing.T)
+QUERY
+  frequency             = 60
+  time_window           = 60
+  auto_mitigate_enabled = %s
```
```suggestion
  auto_mitigation_enabled = %s
```

stephybun reviewed (changes requested):
Hi @cloutier. thanks for the previous changes, unfortunately I was still getting test failures. I've left some comments, once those are fixed up the tests should run and then we can merge this 🙂.

stephybun commented on website/docs/r/monitor_scheduled_query_rules_alert.html.markdown:
```diff
@@ -143,6 +143,8 @@ The following arguments are supported:
 * `trigger` - (Required) The condition that results in the alert rule being run.
 * `action` - (Required) An `action` block as defined below.
 * `authorized_resource_ids` - (Optional) List of Resource IDs referred into query.
+* `auto_mitigate_enabled` - (Optional) Should the alerts in this Metric Alert be auto resolved? Defaults to `true`.
```
```suggestion
* `auto_mitigation_enabled` - (Optional) Should the alerts in this Metric Alert be auto resolved? Defaults to `true`.
```

stephybun commented on website/docs/r/monitor_scheduled_query_rules_alert.html.markdown:
```diff
@@ -143,6 +143,8 @@ The following arguments are supported:
 * `action` - (Required) An `action` block as defined below.
 * `authorized_resource_ids` - (Optional) List of Resource IDs referred into query.
+* `auto_mitigate_enabled` - (Optional) Should the alerts in this Metric Alert be auto resolved? Defaults to `true`.
+-> **NOTE** The `auto_mitigate_enabled` and `throttling` are mutually exclusive and cannot both be set.
```
```suggestion
-> **NOTE** The `auto_mitigation_enabled` and `throttling` are mutually exclusive and cannot both be set.
```

cloutier commented:
@stephybun Thanks for the feedback! :pray: You are helping me getting better at Go and Terraform. Thanks also to @Klaas- for the warning about the pitfalls of this API

I think I'm ready for the final round of test and review

Klaas- reviewed (approved):
I've tested it, it now works and I can switch alerts from suppression to auto mitigation and back, nice work!

stephybun commented:
Thanks @cloutier - tests are passing 🎆. This LGTM 🥳
